src/components/AdminPanelLayout/UsersPane.tsx

```diff
@@ -1,42 +1,156 @@
+      closeModal();
+      setWarningShown(true);
+      if (userToUpdateOnConfirm) {
+        updateUser(userToUpdateOnConfirm[0], userToUpdateOnConfirm[1]);
+      }
+    };
+
+    return (
+      <Layout>
+        <Modal
+          show={showWarning}
+          backdrop={'static'}
+          keyboard={false}
+        >
+          <ModalHeader style={{color: '#fff', backgroundColor: '#d9534f'}}>
+            CAUTION
+          </ModalHeader>
+          <ModalBody>
+            You are editing your own role. Doing so could lead to unexpected behavior, and you may be denied access to the
+            admin panel. Are you certain you wish to proceed?
+          </ModalBody>
+          <ModalFooter>
+            <Button variant={'primary'} onClick={handleCancel}>Cancel</Button>
+            <Button variant={'secondary'} onClick={handleConfirm}>Confirm</Button>
+          </ModalFooter>
+        </Modal>
+      </Layout>
+    );
+  };
+
   return <Layout>{loading ? <HorizontallyCenteredSpinner/> : (
-    <Table striped={true} bordered={true} hover={true}>
-      <thead style={{backgroundColor: '#adb5bd'}}>
-      <tr>
-        <th>Name</th>
-        <th>Email</th>
-        <th>Role</th>
-      </tr>
-      </thead>
-      <tbody>
-      {users.map((u: User) => (
-        <tr key={u.id} style={{cursor: 'pointer'}}>
-          <td>{u.firstName} {u.lastName}</td>
-          <td>{u.email}</td>
-          <td>{u.role}</td>
+    <div>
+      <ConfirmRoleChange/>
+      <div style={{textAlign: 'right', padding: '10px 20px'}}>
+        {commitResponseCode <= 299 && commitResponseCode >= 200 ?
+          (<p style={{position: 'absolute', top: '0', right: '0', color: '#02b875'}}>Save successful!</p>) :
+          commitResponseCode === 401 ?
+            (<p style={{position: 'absolute', top: '0', right: '0', color: '#d9534f'}}>You are not authorized to change
+              user roles, or you are no longer signed in. If you are signed in, and you believe
+              this to be in error, please see a CSLabs admin</p>) :
+            commitResponseCode ?
+              (<p style={{position: 'absolute', top: '0', right: '0', color: '#d9534f'}}>Error saving changes, please try again</p>) :
+              null}
+        <Button variant={'outline-primary'} onClick={commitUsers}>Save</Button>
+      </div>
+      <Table striped={true} bordered={true} hover={true}>
+        <thead style={{backgroundColor: '#adb5bd'}}>
+        <tr>
+          <th>Name</th>
+          <th>Email</th>
+          <th>Role</th>
         </tr>
-      ))}
-      </tbody>
-    </Table>
-  )}</Layout>;
+        </thead>
+        <tbody>
+        {users.map((u: User) => (
+          <UserListItem key={u.id} user={u} onRoleChange={handleRoleChange}/>
+        ))}
+        </tbody>
+      </Table>
+    </div>
+  )}
+  </Layout>;
 
 };
 
-export default UsersPane;
+export default connect((state: WebState) => ({currentUser: getCurrentUser(state) }))(UsersPane);
```

Comment:
I'm having a hard time getting the webstate to update with the response data from here. I think I'm missing something with how this redux function works.

Reply:
I found a solution to get it working. Resolving this.